File: 160-intersection-of-two-linked-lists/160-intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    int getCount(ListNode *head)
    {
        ListNode *cur = head;
        int count = 0;
        while(cur != NULL)
        {
            count++;
            cur = cur -> next;
        }
        return count;
    }
    ListNode *_getIntersectionNode(int d,ListNode *headA,ListNode *headB)
    {
        ListNode *cur1 = headA;
        ListNode *cur2 = headB;
        for(int i=0;i<d;i++)
        {
            if(cur1 == NULL)
            {  
                return NULL;
            }    
            cur1 = cur1->next;
        }
        while(cur1 != NULL && cur2!=NULL)
        {
            if(cur1 == cur2)
                return cur1;
            cur1 = cur1->next;
            cur2 = cur2->next;
        }
        return NULL;
        
        
    }
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        
        int c1 = getCount(headA);
        int c2 = getCount(headB);
        int d;
        
        if(c1 > c2)
        {
            d = c1 - c2;
            return _getIntersectionNode(d,headA,headB);
            
        }
        else
        {
            d = c2 - c1;
            return _getIntersectionNode(d,headB,headA);
        }
        
    }
};

File: 160-intersection-of-two-linked-lists/160-intersection-of-two-linked-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public int getCount(ListNode head)
    {
        ListNode ptr = head;
        int count = 0;
        while(ptr != null )
        {
            count++;
            ptr = ptr.next;
        }
        
        return count;
    }
    public ListNode getIntersection(int d,ListNode headA,ListNode headB)
    {
        ListNode ptr = headA;
        ListNode cpt = headB;
        for(int i=0;i<d;i++)
        {
            if(ptr == null)
                return null;
            ptr = ptr.next;
            
        }
        while(ptr!=null && cpt!=null)
        {
            if(ptr == cpt)
                return ptr;
            ptr = ptr.next;
            cpt = cpt.next;
        }
        return null;
    }
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int c1 = getCount(headA);
        int c2 = getCount(headB);
        int d;
        if(c1 > c2)
        {
            d = c1-c2;
            return getIntersection(d,headA,headB);
        }
        else
        {
            d = c2-c1;
            return getIntersection(d,headB,headA);
        }
        
    }
}